src/util/buffer.rs

```diff
@@ -0,0 +1,266 @@
+//! A memory buffer with resizing.
+
+use core::nonzero::NonZero;
+use std::num::Int;
+use std::ops::Deref;
+use std::rt::heap;
+use std::mem;
+
+/// A safe wrapper around a heap allocated buffer of Ts, tracking capacity only.
+///
+/// Buffer makes no promises about the actual contents of this memory, that's up
+/// to the user of the structure and can be manipulated using the standard pointer
+/// utilities, accessible through the impl of `Deref<Target=*mut T>` for `Buffer<T>`.
+///
+/// As a result of this hands-off approach, `Buffer`s destructor does not attempt
+/// to drop any of the contained elements; the destructor simply frees the contained
+/// memory.
+///
+/// You can think of `Buffer<T>` as an approximation for `Box<[T]>` where the elements
+/// are not guaranteed to be valid/initialized. It is meant to be used as a building
+/// block for other collections, so they do not have to concern themselves with the
+/// minutiae of allocating, reallocating, and deallocating memory.
+pub struct Buffer<T> {
+    buffer: NonZero<*mut T>,
```

Comment:
I would be happy to make this change when it becomes possible, and also change all the underlying allocating utilities to take `NonZero<Unique<T>>` instead of `NonZero<*mut T>`, seems like a really excellent use of using the type system to maintain invariants.